### PR TITLE
[stable/cockroachdb] Complete refactor and capabilities extension

### LINKED ISSUE #18399: [stable/cockroachdb] lowercase letter for variables in values.yaml
The values.yaml should follow the best practices from [https://helm.sh/docs/chart_best_practices/#naming-conventions](https://helm.sh/docs/chart_best_practices/#naming-conventions), so that the variables start with a lowercase letter.

### LINKED ISSUE #18402: [stable/cockroachdb] test pod always fails
**Describe the bug**
If you install the stable/cockroachdb template it starts a test pod. This pod always fails, because it starts before the database is ready.

**Version of Helm and Kubernetes**:
Helm: v2.15.1
Kubernetes: 1.12.9

**Which chart**:
stable/cockroachdb 2.1.15

**What happened**:
Pod has status `Error with exit code 1`:

```
Error: cannot dial server.
Is the server running?
If the server is running, check --host client-side and --advertise server-side.
dial tcp 10.4.7.115:26257: i/o timeout
Failed running "sql"
```

**What you expected to happen**:
Pod has status `Succeeded`.

**How to reproduce it** (as minimally and precisely as possible):
```
helm fetch stable/cockroachdb --version "2.1.15"
helm template "cockroachdb-2.1.15.tgz" | kubectl apply -f -
```

### PR DESCRIPTION
**Pull request by tyranron**
This is a hard backwards-incompatible rework of the `stable/cockroachdb` chart.

Fixes #18399  
Possibly fixes #18402 (at least in all tested cases by me it has passed)

## Motivation

The current `2.x.x` version of `stable/cockroachdb` is hard to reuse in a variety of situations.

Specifically, it bounds you to:
- use only PersistentVolumes for storing data
- use only public-accessible Docker images
- doesn't allow to tweak affinity rules and tolerations

All this made it hard for us to reuse Helm chart both for production/staging/local-dev environments and even numerous E2E tests.

## BC Breaks

The version of the chart with this PR is bumped up to `3.0.0`, reflecting the following backwards-compatibility breaks:
- Values of the chart are fully renamed to follow `camelCase` style (instead of `UpperCase`) widely used in Helm chart's ecosystem (fixes #18399).
    - Additionally, values passed as CLI args to `cockroach` follow the `kebab-case` style to represent a direct mapping from values to these CLI args and omit possible misunderstandings.
- Labels `chart`, `heritage`, `release` and `component` are replaced with more idiomatic `helm.sh/chart`, `app.kubernetes.io/name`, `app.kubernetes.io/instance`, `app.kubernetes.io/managed-by` and `app.kubernetes.io/component`.
- Explicitly differentiate the listening ports in Pods (`conf.port`/`conf.http-port`) and in Services (`service.ports`).
- Remove the `Name` value. Now, the name is idiomatically consists of release name + chart name and can explicitly overwritten with `nameOverride` and `fullnameOverride` values.
- Remove the `Component` value. Now the components of the chart may be differentiated with specifying custom labels for them, if required.

## Added

- Support `nameOverride` and `fullnameOverride` values.
- Ability to specify credentials for pulling used Docker images from private registries.
- Ability to label and annotate almost any chart's resource either dependently and independently (see `values.yml` for details).
- Ability to configure `affinity` rules, `tolerations` taints and `nodeSelector` both for StatefulSet and initialization Job.
- Ability to persist data in:
    - `hostPath` for those who doesn't want to poke (or is not able to) with PersistentVolumes;
    - `emptyDir` which is convenient for temporary snapshot-based deployments (when the container already holds its data and persistence is unnecessary);
    - `persistentVolume` which is still the default one and now supports additional labels and annotations.
- Ability to specify custom ingress rules for NetworkPolicy in addition to existing ones.
- Ability to reuse entrypoint of the used Docker image (previously it was hardly overwritten), which adds the ability to run custom logic before the `cockroach` starts (in our case, for example, this is the data restoration from a snapshot, widely used for numerous E2E tests).
- Ability to deploy in standalone mode (`cockroach start-single-node`) without clustering.
- Ability to configure `logtostderr` verbosity.
- Ability to configure `cluster-name` and `disable-cluster-name-verification`.
- Ability to configure more CLI args as `conf`chart values. While still allowing specify custom CLI extra args (via `statefulset.args`).

## Additionally

From the chart description I've noticed that there are some chart instructions outside, specifically on CockroachDB's site:
> https://www.cockroachlabs.com/docs/v19.2/orchestrate-cockroachdb-with-kubernetes-insecure.html

I'll create PR and update those once (and if) this PR is merged.

## Checklist

- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped (to `3.0.0`)
- [x] All the variables are documented in the `README.md`
- [x] `README.md` info is updated and upgrading from previous versions is described.
- [x] Title of the PR starts with chart name (e.g. `[stable/cockroachdb]`)

### CODE REVIEW

**Comment by tyranron** [comment]
ack @a-robinson @DuskEagle @joshimhoff @keith-mcclellan

File: stable/cockroachdb/templates/_helpers.tpl [hunk 1]
```diff
@@ -1,32 +1,65 @@
+{{/* vim: set filetype=mustache: */}}
```
Comment (DuskEagle) [inline stable/cockroachdb/templates/_helpers.tpl]: Can we remove the vim comment here?

File: stable/cockroachdb/templates/job.init.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
+{{- if gt (.Values.statefulset.replicas | int64) 1 }}
```
Comment (DuskEagle) [inline stable/cockroachdb/templates/job.init.yaml]: I think this should also check if `{{- if .Values.conf.join }}` is set. If a `join` flag is passed to the `cockroach start` command in the statefulset, then this job is required, even when only running one CRDB pod.

**Comment by tyranron** [comment]
Update:
- Remove vim comment from `templates/_helpers.tpl`.
- Run init Job when `conf.join` values is set.
- Fix mounting certs in init Job when `tls.enabled: no`.
- Make `restartPolicy` of init Job be always `OnFailure` (not just on `tls.enabled: yes`), so omit warnings like "incorrect value 'Always' for 'restartPolicy'".
- Merge with latest `master` and `stable/cockroachdb` chart updates.

ack @DuskEagle

**Comment by DuskEagle** [comment]
Hi @tyranron, thanks for taking the time for this huge refactor. Overall, I think this is a big improvement, and I'm excited to see it get in.

I've been doing some testing locally and creating CockroachDB clusters has worked exactly as expected. I did however run into a couple of issues when changing the number of replicas from 1 to 3 on an existing deployment:

* Because the cockroachdb-init job did not run when there was only one replica, when we change the number of replicas, the job gets created and attempts to initialize an already-initalized cluster. This results in the job [running forever](https://github.com/helm/charts/blob/688b97dac864d7fb1b8f2907ae5ada319d17cf66/stable/cockroachdb/templates/job.init.yaml#L94-L106), re-attempting to initialize the cluster every five seconds.
* When changing the number of replicas, we change the `--join` flag in the StatefulSet spec. This causes the existing pod to restart once the two new pods are ready. This will actually happen whenever you scale up the number of replicas, which is unfortunate. But it's especially bad when moving from 1 to 3 replicas, because the existing ranges have not been replicated over to the new pods yet. This causes the new pods to crash until the `cockroachdb-0` pod has restarted.

For the first issue, it might be simplest just to always pass a `--join` flag to `cockroach start`, even for a one-node cluster. Alternatively, we could fix the `cockroachdb-init` job to exit successfully if it is attempting to init an already-initialized cluster. Both approaches seem good to me.

The second issue is a bit trickier. I think this problem is actually one of the reasons why we didn't change the `--join` flag before. In theory, we could set the `.spec.updateStrategy.type` to [`OnDelete`](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete) to ensure that existing pods aren't restarted. But we really only want that update strategy when changing the number of replicas, not when changing other parts of the statefulset spec (such as the CockroachDB version).

As long as the join flag includes just one running cockroachdb pod, then new pods will be able to join the cluster and gossip with all existing pods. So the join flag only has to contain enough pods so that you can still scale even if some pods are down for whatever reason (such as an AZ outage).

So, the simplest approach might be to continue to just hard-code the join flag to three pods (even for a one-pod cluster, this is harmless). Alternatively, we could include instructions to always set the `updateStrategy` to `OnDelete` when changing the number of replicas and to set the `updateStrategy` to `RollingUpdate` when performing any other kind of update, but that might be harder for users of the chart.

**Comment by tyranron** [comment]
@DuskEagle thank you for so detailed feedback and giving info tips about how CockroachDB clustering works.

> Alternatively, we could fix the cockroachdb-init job to exit successfully if it is attempting to init an already-initialized cluster. Both approaches seem good to me.

Actually, the idempotence of init Job is quite a good thing despite the issues above. I believe this should be first-class supported in `/cockroach init` command, but at the moment I've solved with some hairy bash magic.

> So, the simplest approach might be to continue to just hard-code the join flag to three pods (even for a one-pod cluster, this is harmless). 

I didn't know it's harmless. So now, after you've described the bigger picture I see that 3 hardcoded Pods is the best solution in this situation without introducing Kubernetes Operators in the game.

I've documented reasoning about hardcoded Pods for future contributors.

--------------------------------

Update:
- Hardcode `--join` Pods always to `3` (if the flag is not specified explicitly).
- Make init Pod idempotent (succeeds if luster has been initialized already).
    - Set `terminationGracePeriodSeconds: 0` for init Job as it doesn't handle OS signals anyway.
- Rebased onto latest `master`.

**Comment by tyranron** [comment]
@DuskEagle after investigating the `cockroach start-single-node` command all the described above makes even more sense.

So, with latest update I've introduced `conf.single-node` value (defaults to `no`), which gives the ability to run CockroachDB in a standalone mode without requiring initialization (so no init Job). This would be quite useful for some temporary non-production deployments (like E2E tests, etc). I've documented the value and placed the explicit warning on it about possible consequences.

----------------------------------

Update:
- Provide `conf.single-node` value which deploys Pods in standalone mode without initialization Pod.
    - [x] value documented in `values.yaml`;
    - [x] README updated.
- Support `conf.cluster-name` and `conf.disable-cluster-name-verification` values.
    - [x] value documented in `values.yaml`;
    - [x] README updated.
- Return printing of executed command in initialization Pod.
- Remove `--port` flag usage in initialization Pod and add port to `--host` flag directly.
- Rebase onto latest `master`.

**Comment by DuskEagle** [comment]
Thanks for the update! I've tested scaling clusters and running the init job against an already-initialized cluster, and both now work exactly as expected. I also tested starting a cluster using the `single-node` helm value and that worked as well.

> So now, after you've described the bigger picture I see that 3 hardcoded Pods is the best solution in this situation without introducing Kubernetes Operators in the game.

Introducing a public operator is something we've been talking about internally. We think it could make several common operations simpler, such as creating or scaling secure clusters, or upgrading between major versions of CockroachDB. If you have any feedback you'd want to give on this please feel free to bring it up on our [Github](https://github.com/cockroachdb/cockroach), [Forums](https://forum.cockroachlabs.com/), or [Slack](https://cockroa.ch/slack).

Thanks again for all these improvements to this Helm chart!

/lgtm /approve

**Pull request merged**